File: app/src/main/java/com/sasaj/graphics/drawingapp/ui/authentication/RegisterActivity.java
package com.sasaj.graphics.drawingapp.ui.authentication;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;

import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser;
import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUserAttributes;
import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUserCodeDeliveryDetails;
import com.amazonaws.mobileconnectors.cognitoidentityprovider.handlers.SignUpHandler;
import com.sasaj.graphics.drawingapp.R;
import com.sasaj.graphics.drawingapp.cognito.CognitoHelper;
import com.sasaj.graphics.drawingapp.ui.main.MainActivity;

public class RegisterActivity extends AppCompatActivity {

    private static final String TAG = RegisterActivity.class.getSimpleName();
    private AutoCompleteTextView username;
    private AutoCompleteTextView email;
    private EditText password;
    private Button signupButton;

    private final CognitoUserAttributes attributes = new CognitoUserAttributes();

    private final SignUpHandler signUpHandler = new SignUpHandler() {
        @Override
        public void onSuccess(CognitoUser user, boolean signUpConfirmationState, CognitoUserCodeDeliveryDetails cognitoUserCodeDeliveryDetails) {
            Log.i(TAG, "onSuccess - is confirmed: " + signUpConfirmationState);
            if (!signUpConfirmationState) {
                Log.i(TAG, "onSuccess: not confirmed, token sent to: " + cognitoUserCodeDeliveryDetails.getDestination());
                Intent intent =  new Intent(RegisterActivity.this, VerifyActivity.class);
                startActivity(intent);
            } else {
                Log.i(TAG, "onSuccess - confirmed: ");
                Intent intent =  new Intent(RegisterActivity.this, MainActivity.class);
                startActivity(intent);
            }
        }

        @Override
        public void onFailure(Exception exception) {
            Log.e(TAG, "Signup Failure: ", exception);
        }
    };


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        username = findViewById(R.id.username);
        email = findViewById(R.id.email);
        password = findViewById(R.id.password);
        signupButton = findViewById(R.id.sign_up_button);

        signupButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                attributes.addAttribute("email", email.getText().toString());

                CognitoHelper cognitoHelper = new CognitoHelper(RegisterActivity.this);
                cognitoHelper.getUserPool().signUpInBackground(username.getText().toString(),
                                                                  password.getText().toString(),
                                                                  attributes,
                                                                  null,
                                                                  signUpHandler);

            }
        });
    }
}

File: app/src/main/java/com/sasaj/graphics/drawingapp/ui/authentication/RegisterActivity.kt
package com.sasaj.graphics.drawingapp.ui.authentication

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.util.Log
import com.sasaj.graphics.drawingapp.R
import com.sasaj.graphics.drawingapp.ui.base.BaseActivity
import com.sasaj.graphics.drawingapp.ui.main.MainActivity
import com.sasaj.graphics.drawingapp.viewmodel.RegisterViewModel
import com.sasaj.graphics.drawingapp.viewmodel.common.Response
import com.sasaj.graphics.drawingapp.viewmodel.common.Status
import kotlinx.android.synthetic.main.activity_register.*

class RegisterActivity : BaseActivity() {

    private lateinit var vm: RegisterViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        vm = ViewModelProviders.of(this).get(RegisterViewModel::class.java)
        vm.getRegisterLiveData().observe(this, Observer { response -> processResponse(response) })

        signUpButton!!.setOnClickListener {
            val attributes: HashMap<String, String> = HashMap()
            attributes.put("email", email!!.text.toString())
            vm.register(username!!.text.toString(), password!!.text.toString(), attributes)

        }
    }

    private fun processResponse(response: Response?) {
        when (response?.status) {
            Status.LOADING -> renderLoadingState()
            Status.SUCCESS -> renderSucessRegisteringState(response.data)
            Status.ERROR -> renderErrorState(response.error)
        }
    }

    override fun resetViewModel() {
        vm.resetLiveData()
    }

    private fun renderLoadingState() {
        showProgress("wait...")
    }

    private fun renderSucessRegisteringState(status: String?) {
        hideProgress()
        if (status == "confirmed") {
            Log.i(TAG, "Succesfully registered")
            val intent = Intent(this@RegisterActivity, MainActivity::class.java)
            startActivity(intent)
            finish()
        } else {
            Log.i(TAG, "Go to verify")
            val intent = Intent(this@RegisterActivity, VerifyActivity::class.java)
            startActivity(intent)
            finish()
        }
    }

    private fun renderErrorState(throwable: Throwable?) {
        hideProgress()
        Log.e(TAG, "Error logging in ", throwable)
        showDialogMessage("Error logging in", throwable.toString())
    }


    companion object {

        private val TAG = RegisterActivity::class.java.simpleName
    }
}
